File: llm-wiki/src/lib/page-blocks.ts
export type PageHeadingBlock = {
  id: string
  type: 'heading'
  level: number
  text: string
}

export type PageParagraphBlock = {
  id: string
  type: 'paragraph' | 'quote'
  text: string
}

export type PageBulletListBlock = {
  id: string
  type: 'bullet_list'
  items: string[]
}

export type PageTodoItem = {
  text: string
  checked: boolean
}

export type PageTodoListBlock = {
  id: string
  type: 'todo_list'
  items: PageTodoItem[]
}

export type PageImageBlock = {
  id: string
  type: 'image'
  url: string
  caption?: string
  alt?: string
  assetId?: string | null
}

export type PageBlock =
  | PageHeadingBlock
  | PageParagraphBlock
  | PageBulletListBlock
  | PageTodoListBlock
  | PageImageBlock

const IMAGE_RE = /^!\[([^\]]*)\]\(([^)]+)\)\s*$/
const HEADING_RE = /^(#{1,6})\s+(.+?)\s*$/
const TODO_RE = /^\s*[-*]\s+\[([xX ])\]\s+(.+?)\s*$/
const BULLET_RE = /^\s*[-*+]\s+(.+?)\s*$/

function blockId() {
  return `blk-${Math.random().toString(36).slice(2, 10)}`
}

export function createParagraphBlock(text = ''): PageParagraphBlock {
  return { id: blockId(), type: 'paragraph', text }
}

export function createHeadingBlock(text = 'Untitled section', level = 2): PageHeadingBlock {
  return { id: blockId(), type: 'heading', level, text }
}

export function createImageBlock(url: string, caption = '', assetId?: string | null): PageImageBlock {
  return { id: blockId(), type: 'image', url, caption, alt: caption, assetId }
}

export function normalizePageBlocks(input?: unknown): PageBlock[] {
  if (!Array.isArray(input) || input.length === 0) return [createParagraphBlock('')]
  const blocks = input.flatMap<PageBlock>((raw) => {
    if (!raw || typeof raw !== 'object') return []
    const block = raw as Record<string, unknown>
    const type = String(block.type || 'paragraph')
    const id = String(block.id || blockId())
    if (type === 'heading') {
      return [{ id, type: 'heading', level: Math.min(Math.max(Number(block.level || 1), 1), 6), text: String(block.text || '') }]
    }
    if (type === 'image') {
      const url = String(block.url || '')
      if (!url) return []
      return [{ id, type: 'image', url, caption: String(block.caption || ''), alt: String(block.alt || block.caption || ''), assetId: block.assetId ? String(block.assetId) : null }]
    }
    if (type === 'bullet_list') {
      const items = Array.isArray(block.items) ? block.items.map(item => String(item).trim()) : ['']
      return [{ id, type: 'bullet_list', items: items.length ? items : [''] }]
    }
    if (type === 'todo_list') {
      const items = Array.isArray(block.items)
        ? block.items.map((item) => {
            const record = item as Record<string, unknown>
            return { text: String(record?.text || ''), checked: Boolean(record?.checked) }
          })
        : [{ text: '', checked: false }]
      return [{ id, type: 'todo_list', items: items.length ? items : [{ text: '', checked: false }] }]
    }
    if (type === 'quote') {
      return [{ id, type: 'quote', text: String(block.text || '') }]
    }
    return [{ id, type: 'paragraph', text: String(block.text || '') }]
  })
  return blocks.length ? blocks : [createParagraphBlock('')]
}

export function markdownToPageBlocks(markdown?: string): PageBlock[] {
  const content = (markdown || '').trim()
  if (!content) return [createParagraphBlock('')]

  const blocks: PageBlock[] = []
  let paragraphLines: string[] = []
  let bulletItems: string[] = []
  let todoItems: PageTodoItem[] = []
  let quoteLines: string[] = []

  const flushParagraph = () => {
    if (paragraphLines.length) {
      blocks.push(createParagraphBlock(paragraphLines.join('\n')))
      paragraphLines = []
    }
  }
  const flushBullets = () => {
    if (bulletItems.length) {
      blocks.push({ id: blockId(), type: 'bullet_list', items: [...bulletItems] })
      bulletItems = []
    }
  }
  const flushTodos = () => {
    if (todoItems.length) {
      blocks.push({ id: blockId(), type: 'todo_list', items: [...todoItems] })
      todoItems = []
    }
  }
  const flushQuotes = () => {
    if (quoteLines.length) {
      blocks.push({ id: blockId(), type: 'quote', text: quoteLines.join('\n') })
      quoteLines = []
    }
  }
  const flushAll = () => {
    flushParagraph()
    flushBullets()
    flushTodos()
    flushQuotes()
  }

  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trimEnd()
    const trimmed = line.trim()
    if (!trimmed) {
      flushAll()
      continue
    }
    const imageMatch = trimmed.match(IMAGE_RE)
    if (imageMatch) {
      flushAll()
      blocks.push(createImageBlock(imageMatch[2], imageMatch[1]))
      continue
    }
    const headingMatch = trimmed.match(HEADING_RE)
    if (headingMatch) {
      flushAll()
      blocks.push(createHeadingBlock(headingMatch[2], headingMatch[1].length))
      continue
    }
    const todoMatch = trimmed.match(TODO_RE)
    if (todoMatch) {
      flushParagraph()
      flushBullets()
      flushQuotes()
      todoItems.push({ text: todoMatch[2], checked: todoMatch[1].toLowerCase() === 'x' })
      continue
    }
    const bulletMatch = trimmed.match(BULLET_RE)
    if (bulletMatch) {
      flushParagraph()
      flushTodos()
      flushQuotes()
      bulletItems.push(bulletMatch[1])
      continue
    }
    if (trimmed.startsWith('>')) {
      flushParagraph()
      flushBullets()
      flushTodos()
      quoteLines.push(trimmed.replace(/^>\s?/, ''))
      continue
    }
    flushBullets()
    flushTodos()
    flushQuotes()
    paragraphLines.push(line)
  }

  flushAll()
  return blocks.length ? blocks : [createParagraphBlock('')]
}

export function pageBlocksToMarkdown(blocks: PageBlock[]): string {
  const normalized = normalizePageBlocks(blocks)
  const lines: string[] = []
  normalized.forEach((block) => {
    if (block.type === 'heading') {
      lines.push(`${'#'.repeat(block.level)} ${block.text}`.trimEnd())
    } else if (block.type === 'image') {
      lines.push(`![${(block.caption || block.alt || 'image').trim()}](${block.url})`)
    } else if (block.type === 'bullet_list') {
      block.items.forEach(item => lines.push(`- ${item}`.trimEnd()))
    } else if (block.type === 'todo_list') {
      block.items.forEach(item => lines.push(`- [${item.checked ? 'x' : ' '}] ${item.text}`.trimEnd()))
    } else if (block.type === 'quote') {
      block.text.split('\n').forEach(line => lines.push(`> ${line}`.trimEnd()))
    } else {
      lines.push(block.text.trimEnd())
    }
    lines.push('')
  })
  return lines.join('\n').trim()
}

File: backend/app/services/page_blocks.py
from __future__ import annotations

import re
from copy import deepcopy
from uuid import uuid4

IMAGE_RE = re.compile(r"^!\[(?P<alt>[^\]]*)\]\((?P<url>[^)]+)\)\s*$")
HEADING_RE = re.compile(r"^(?P<hashes>#{1,6})\s+(?P<text>.+?)\s*$")
TODO_RE = re.compile(r"^\s*[-*]\s+\[(?P<checked>[xX ])\]\s+(?P<text>.+?)\s*$")
BULLET_RE = re.compile(r"^\s*[-*+]\s+(?P<text>.+?)\s*$")


def _block_id() -> str:
    return f"blk-{uuid4().hex[:10]}"


def _text_block(block_type: str, text: str, **extra) -> dict:
    return {"id": _block_id(), "type": block_type, "text": text.strip(), **extra}


def markdown_to_blocks(content_md: str | None) -> list[dict]:
    content = (content_md or "").strip()
    if not content:
        return [{"id": _block_id(), "type": "paragraph", "text": ""}]

    blocks: list[dict] = []
    paragraph_lines: list[str] = []
    bullet_items: list[str] = []
    todo_items: list[dict] = []
    quote_lines: list[str] = []

    def flush_paragraph():
        nonlocal paragraph_lines
        if paragraph_lines:
            blocks.append(_text_block("paragraph", "\n".join(paragraph_lines)))
            paragraph_lines = []

    def flush_bullets():
        nonlocal bullet_items
        if bullet_items:
            blocks.append({"id": _block_id(), "type": "bullet_list", "items": bullet_items[:]})
            bullet_items = []

    def flush_todos():
        nonlocal todo_items
        if todo_items:
            blocks.append({"id": _block_id(), "type": "todo_list", "items": deepcopy(todo_items)})
            todo_items = []

    def flush_quotes():
        nonlocal quote_lines
        if quote_lines:
            blocks.append(_text_block("quote", "\n".join(quote_lines)))
            quote_lines = []

    def flush_all():
        flush_paragraph()
        flush_bullets()
        flush_todos()
        flush_quotes()

    for raw_line in content.splitlines():
        line = raw_line.rstrip()
        stripped = line.strip()
        if not stripped:
            flush_all()
            continue

        image_match = IMAGE_RE.match(stripped)
        if image_match:
            flush_all()
            alt = image_match.group("alt").strip()
            url = image_match.group("url").strip()
            blocks.append(
                {
                    "id": _block_id(),
                    "type": "image",
                    "url": url,
                    "caption": alt,
                    "alt": alt,
                }
            )
            continue

        heading_match = HEADING_RE.match(stripped)
        if heading_match:
            flush_all()
            blocks.append(
                {
                    "id": _block_id(),
                    "type": "heading",
                    "level": len(heading_match.group("hashes")),
                    "text": heading_match.group("text").strip(),
                }
            )
            continue

        todo_match = TODO_RE.match(stripped)
        if todo_match:
            flush_paragraph()
            flush_bullets()
            flush_quotes()
            todo_items.append(
                {
                    "text": todo_match.group("text").strip(),
                    "checked": todo_match.group("checked").lower() == "x",
                }
            )
            continue

        bullet_match = BULLET_RE.match(stripped)
        if bullet_match:
            flush_paragraph()
            flush_todos()
            flush_quotes()
            bullet_items.append(bullet_match.group("text").strip())
            continue

        if stripped.startswith(">"):
            flush_paragraph()
            flush_bullets()
            flush_todos()
            quote_lines.append(stripped.removeprefix(">").strip())
            continue

        flush_bullets()
        flush_todos()
        flush_quotes()
        paragraph_lines.append(line)

    flush_all()
    return blocks or [{"id": _block_id(), "type": "paragraph", "text": ""}]


def sanitize_blocks(content_json: list[dict] | None) -> list[dict]:
    if not isinstance(content_json, list):
        return [{"id": _block_id(), "type": "paragraph", "text": ""}]

    sanitized: list[dict] = []
    for raw in content_json:
        if not isinstance(raw, dict):
            continue
        block_type = str(raw.get("type") or "paragraph")
        block_id = str(raw.get("id") or _block_id())
        if block_type == "heading":
            text = str(raw.get("text") or "").strip()
            level = int(raw.get("level") or 1)
            sanitized.append({"id": block_id, "type": "heading", "level": min(max(level, 1), 6), "text": text})
        elif block_type == "image":
            url = str(raw.get("url") or "").strip()
            if not url:
                continue
            caption = str(raw.get("caption") or "").strip()
            alt = str(raw.get("alt") or caption).strip()
            asset_id = raw.get("assetId")
            sanitized.append({"id": block_id, "type": "image", "url": url, "caption": caption, "alt": alt, "assetId": asset_id})
        elif block_type == "bullet_list":
            items = [str(item).strip() for item in raw.get("items") or [] if str(item).strip()]
            sanitized.append({"id": block_id, "type": "bullet_list", "items": items or [""]})
        elif block_type == "todo_list":
            items = []
            for item in raw.get("items") or []:
                if isinstance(item, dict):
                    text = str(item.get("text") or "").strip()
                    items.append({"text": text, "checked": bool(item.get("checked"))})
            sanitized.append({"id": block_id, "type": "todo_list", "items": items or [{"text": "", "checked": False}]})
        elif block_type == "quote":
            sanitized.append({"id": block_id, "type": "quote", "text": str(raw.get("text") or "").strip()})
        else:
            sanitized.append({"id": block_id, "type": "paragraph", "text": str(raw.get("text") or "").strip()})

    return sanitized or [{"id": _block_id(), "type": "paragraph", "text": ""}]


def blocks_to_markdown(content_json: list[dict] | None) -> str:
    blocks = sanitize_blocks(content_json)
    lines: list[str] = []
    for block in blocks:
        block_type = block.get("type")
        if block_type == "heading":
            lines.append(f"{'#' * int(block.get('level', 1))} {str(block.get('text') or '').strip()}".rstrip())
        elif block_type == "image":
            alt = str(block.get("caption") or block.get("alt") or "image").strip()
            lines.append(f"![{alt}]({str(block.get('url') or '').strip()})")
        elif block_type == "bullet_list":
            lines.extend(f"- {item}" for item in (block.get("items") or []))
        elif block_type == "todo_list":
            lines.extend(f"- [{'x' if item.get('checked') else ' '}] {item.get('text', '').strip()}".rstrip() for item in (block.get("items") or []))
        elif block_type == "quote":
            quote_text = str(block.get("text") or "").splitlines() or [""]
            lines.extend(f"> {line}".rstrip() for line in quote_text)
        else:
            lines.append(str(block.get("text") or "").rstrip())
        lines.append("")
    return "\n".join(lines).strip()


def normalize_page_document(content_md: str | None, content_json: list[dict] | None) -> tuple[str, list[dict]]:
    if content_json:
        normalized_blocks = sanitize_blocks(content_json)
        normalized_md = (content_md or "").strip() or blocks_to_markdown(normalized_blocks)
        return normalized_md, normalized_blocks
    normalized_md = (content_md or "").strip()
    normalized_blocks = markdown_to_blocks(normalized_md)
    return normalized_md, normalized_blocks
